### PR TITLE
Sparkmagic, autovizwidget, and hdijupyterutils

### PR DESCRIPTION
**Pull request by aggFTW**
Separates into three different packages:

1. sparkmagic: kernels for spark through livy client 
2. autovizwidget: ipywidget that inspects a pandas df and does visualization with plotly
3. hdijupyterutils: common code shared by the first two packages

Apart from putting everything into different folders, tackled this:
- Make the three packages ready to submit to PyPI
- Separate configs for sparkmagic and autovizwidget
- Separate constants for every package
- Separate event emitters for different packages

### CODE REVIEW

File: autovizwidget/autovizwidget/plotlygraphs/graphrenderer.py [hunk 1]
```diff
@@ -10,20 +10,14 @@
...
-from remotespark.datawidgets.encoding import Encoding
+from ..widget.encoding import Encoding
 
 
 class GraphRenderer(object):
 
     @staticmethod
     def render(df, encoding, output):
         with output:
-            # HACK HACK HACK!
```
Comment (msftristew) [inline autovizwidget/autovizwidget/plotlygraphs/graphrenderer.py]: This was removed why?

File: autovizwidget/autovizwidget/utils/configuration.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+from .constants import HOME_PATH, CONFIG_FILE
+    
+
+d = {}
+path = join_paths(HOME_PATH, CONFIG_FILE)
+
+
+def override(config, value):
+    global d, path
```
Comment (msftristew) [inline autovizwidget/autovizwidget/utils/configuration.py]: Are these `global` declarations necessary?  I don't think so

File: autovizwidget/autovizwidget/utils/configuration.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def override_all(obj):
+    global d
+    _override_all(d, obj)
+    
+    
+# Configs
+
+@with_override(d, path)
```
Comment (msftristew) [inline autovizwidget/autovizwidget/utils/configuration.py]: Is there a reason you have to call `with_override(d, path)` every time?  Can't you just say

    _with_override = with_override(d, path)

up top and then use the `@_with_override` decorator on all the configurations?

File: autovizwidget/autovizwidget/utils/configuration.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Distributed under the terms of the Modified BSD License.
+from hdijupyterutils.constants import EVENTS_HANDLER_CLASS_NAME, LOGGING_CONFIG_CLASS_NAME
+from hdijupyterutils.utils import join_paths
+from hdijupyterutils.configuration import override as _override
```
Comment (msftristew) [inline autovizwidget/autovizwidget/utils/configuration.py]: I don't like this API at all.  Don't make any changes for now, but let's rethink it and address it in a later PR :)

File: hdijupyterutils/hdijupyterutils/configuration.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    line = u"".join(config_text).strip()
+    
+    if line == u"":
+        overrides = {}
+    else:
+        overrides = json.loads(line)
+    return overrides
+
+
+def logging_config():
```
Comment (msftristew) [inline hdijupyterutils/hdijupyterutils/configuration.py]: Should go in `log.py`?

File: hdijupyterutils/hdijupyterutils/utils.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        first_run = False
+        instance_id = generate_uuid()
+
+    if instance_id is None:
+        raise ValueError("Tried to return empty instance ID.")
+
+    return instance_id
+
+
+def coerce_pandas_df_to_numeric_datetime(df):
```
Comment (msftristew) [inline hdijupyterutils/hdijupyterutils/utils.py]: This probably shouldn't be here -- should be either in autovizwidget or sparkmagic depending on where it's used

File: sparkmagic/sparkmagic/utils/configuration.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+
+# Configs
+       
+def get_session_properties(language):
+    properties = copy.deepcopy(session_configs())
+    properties[u"kind"] = get_livy_kind(language)
+    return properties
+
+@with_override(d, path)
```
Comment (msftristew) [inline sparkmagic/sparkmagic/utils/configuration.py]: Same comment as above: is calling `with_override(d, path)` every time necessary?

File: autovizwidget/autovizwidget/plotlygraphs/graphrenderer.py (hunk 1 above)
Comment (aggFTW) [inline autovizwidget/autovizwidget/plotlygraphs/graphrenderer.py]: They fixed it on plotly 1.10

File: autovizwidget/autovizwidget/utils/configuration.py (hunk 2 above)
Comment (aggFTW) [inline autovizwidget/autovizwidget/utils/configuration.py]: It is, otherwise it looks for it locally.

File: autovizwidget/setup.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        'Environment :: Console',
+        'Intended Audience :: Science/Research',
+        'License :: OSI Approved :: BSD License',
+        'Natural Language :: English',
+        'Programming Language :: Python :: 2.6',
+        'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3.3',
+        'Programming Language :: Python :: 3.4'],
+      install_requires=[
+          'plotly==1.10.0',
```
Comment (msftristew) [inline autovizwidget/setup.py]: This is a really restrictive requirement, make it `plotly>=1.10.0` or `plotly>=1.10.0,<1.11.0`

Comment (msftristew) [inline autovizwidget/setup.py]: Or `plotly>=1.10.0,<2`

File: autovizwidget/autovizwidget/utils/configuration.py (hunk 2 above)
Comment (msftristew) [inline autovizwidget/autovizwidget/utils/configuration.py]: Just tried this in a new Python script:

    d = {}
    path = 'hi'

    def override():
        print d, path

    override()

... and it works.

Comment (aggFTW) [inline autovizwidget/autovizwidget/utils/configuration.py]: mmm changing it... I must have been doing something different.

File: autovizwidget/setup.py (hunk 8 above)
Comment (aggFTW) [inline autovizwidget/setup.py]: I think we'll relax that once we see plotly stabilize. From what we've seen, 1.11 could end up breaking us.

Comment (msftristew) [inline autovizwidget/setup.py]: `plotly>=1.10.0,<1.11.0` is fine in that case.  Please don't leave it as `==1.10.0`, that's extremely restrictive :)

**Comment by msftristew** [comment]
LGTM

File: sparkmagic/README.md [hunk 9]
```diff
@@ -32,7 +32,7 @@ Check out the [examples](examples) directory.
         jupyter-kernelspec install sparkmagic/kernels/sparkkernel
         jupyter-kernelspec install sparkmagic/kernels/pysparkkernel
         
-3. (Optional) Modify the configuration file at ~/.sparkmagic/config.json. Look at the example config at https://github.com/jupyter-incubator/hdijupyterutils
+3. (Optional) Modify the configuration file at ~/.sparkmagic/config.json. Look at the `example_config.json`
```
Comment (msftristew) [inline sparkmagic/README.md]: Include a link

File: sparkmagic/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
+# sparkmagic
+
+Sparkmagic is a set of tools for interactively working with remote Spark clusters through [Livy](https://github.com/cloudera/hue/tree/master/apps/spark/java), a Spark REST server, in [Jupyter](http://jupyter.org) notebooks.
+The Sparkmagic project includes a set of magics for interactively running Spark code in multiple languages, as well as some kernels that you can use to turn Jupyter into an integrated Spark environment.
+
+![Automatic SparkContext and SQLContext creation](screenshots/sparkcontext.png)
```
Comment (msftristew) [inline sparkmagic/README.md]: Broken links

File: sparkmagic/README.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
+# sparkmagic
```
Comment (msftristew) [inline sparkmagic/README.md]: Should delete this file since a copy is in the root of this repo?

**Pull request merged**